pandas/util/testing.py

```diff
@@ -1415,27 +1440,36 @@ def assert_equal(left, right, **kwargs):
     """
     __tracebackhide__ = True
 
-    if isinstance(left, pd.Index):
+    if isinstance(left, Index):
+        assert isinstance(right, Index)
```

Comment:
What is the message here? These changes also impact the tests so don't think we want these

Reply:
yeah we don't want to actually change the semantics here, remove these

Reply:
If I comment out `assert isinstance(right, Index)` then mypy complains about `assert_index_equal(left, right, **kwargs)`:
```
pandas/util/testing.py:1443: error: Argument 2 to "assert_index_equal" has incompatible type "Union[DataFrame, Index]"; expected "Index"
pandas/util/testing.py:1443: error: Argument 2 to "assert_index_equal" has incompatible type "Union[DataFrame, Any]"; expected "Index"
```
But assuming left is an index, assert_index_equal(left, right) should raise if right isn't an index, so I don't think adding an `assert isinstance(right, index)` changes the semantics...? We will just catch the exception earlier based on the type. 
I got the idea from @simonjayhawkins but I might have misunderstood. 
Happy to do something different if there's a better way of handling this.

Reply:
OK I've just pushed another attempt (0c2f6926f0de23e26ff8983ff14100672a01e99b) that uses the `_check_isinstance(left, right, Index)` command, which is always called by assert_index_equal. There's no change in semantics as far as I can tell. If that line doesn't raise then it's safe to case right to Index which keeps mypy happy. 
If this is OK I can try and do something similar for the other types in assert_equal. Please let me know what you think.